### ISSUE TITLE
Add nuance to how stale works to encourage interaction from Maintainers in situations where contributors follow all the guidelines

### ISSUE THREAD

**Issue by anglinb** [probot/stale#67]
👋 Wanted to memorialize a conversation that @gr2m @bkeepers @JasonEtco and I had with @keithamus at GitHub Universe this past week. Keith brought up a really valid point about first time contributors and the way a bot closing out their issue might affect future contributions to open source. **Everyone mentioned above, please feel free to edit my explanation as you see fit.**

Apologies for my paraphrasing: There are two types of issues that end up getting closed. There are issues where the submitter has done their best to reproduce and for one reason or another the issue doesn't get attention. The type is when the submitter is not responding or might not be following instructions. This type makes perfect sense for stale. It will close out issues that don't meet the bar for human interaction. 

However, the first type of issues where the submitter has done everything in their power to help describe the issue, there might be an opportunity to change the tone or interaction so that we could help empower new comers rather than have them get shut out by a bot. 

### Examples
I tried to pull up a few issues to try and describe these different circumstances. 

Here is a situation where a person opened an issue without filling out the proper documentation: https://github.com/atom/atom/issues/15889 This is a perfect candidate for stale or another bot to reply as rsese did in this instance. 

Here is a situation when the original opener of the issue `styurin` got an answer from a member of the project and never replied. In this case it seems that it's reasonable to close the issue b/c the original submitter never followed up: https://github.com/atom/atom/issues/12964

Here is a situation when the user took time to report an issue and someone added a label but there was no further interaction. I would imagine the issue got fixed at some point b/c the original submitter, `misaghshakeri`, never replied. https://github.com/atom/atom/issues/12548 However, I wonder if some sort of interaction or nuance could help provide context. 

### Possible Solutions
We could use the first-timer property returned from the GraphQL endpoint to change the behavior of the app. 

We could build a heuristic for how new a person is to GitHub to provide a nuanced response to stale.

We could be more opinionated/prescriptive on how we ask users to write their stale comment text. 
---

In short, I think stale is awesome. It's adoption and resulting data gives us a really cool opportunity to re-asses its interactions and further investigate the tone, context, and persona of stale that will best serve the needs of first time contributors up through maintainers.

**Comment by keithamus** [comment]
This seems like a great summary of our conversation! Thanks for putting this down @anglinb 😄 💯

So yeah - the take-away point is that issues can become "stale" for many different reasons; and there's more to a stale issue than always warrants unilateral closing. I believe the first-pass of stale-bot interacting with an issue should be prompting both the user and the maintainers for any more action - even if the user has still filled out a perfect issue report, they may be no longer experiencing the issue or could provide new output to further diagnose the issue; alternatively maintainers may not have gotten round to responding properly yet and just need nudging. I understand the message is configurable so it's just down to the team - but I agree with the solution of making this more opinionated or at least offering guidance of writing a friendly+helpful first-response could help.

Also - I think the simple action of a bot automatically closing the issue could be deemed user hostile. _Perhaps_ rather than the bot closing the issue, it could be configurable that the bot only comments on stale issues, and leaves it to the maintainers to decide if it should be closed? Having a message every N days saying "hey, this issue has gone [another] N days without any activity, someone should do something" could be a great way to prompt either the user or maintainer to use human-intuition to drive the issue to close/resolved. I believe (though would be happily corrected) that currently the path is `stale->closed` or `stale->updated->stale->closed` - having a loop of `stale->reminder->reminder->reminder->...` may be beneficial.

Another thing is; I believe that in some ways `stale` draws negative connotations. I wonder if there is a better word to use to describe this behaviour? maybe `needs-attention` over `stale`?

**Comment by gr2m** [comment]
I’d love stale bot to become more nuanced, thanks for putting such a great issue together to get the discussion going.

I wouldn’t mind an option to let humans close the issue, but I think it’s still okay for bots to make that by default. In most cases it is just being honest. Maintainers would love to help out more but they simply can’t. Even if the issues are relevant and pull requests are good, if they don’t get merged or reviewed after a certain time, it’s just being honest to close them, because it won’t happen any time soon.

But we can make the explanations more helpful. It would also be a great opportunity to give maintainers some guidance on what options they have, how people tend to react, what pitfalls to avoid, etc.

**Comment by Ben3eeE** [comment]
After reading some of the reactions we got we updated the stale comment to be more descriptive /cc: https://github.com/atom/atom/pull/15775.

Here are some examples of reactions to the bot. I can add more as I find them.

https://github.com/atom/atom/issues/1495#issuecomment-178383290
https://github.com/atom/atom/issues/11733 / https://github.com/atom/atom/issues/12677
https://github.com/atom/atom/issues/5998

**Comment by mezz** [comment]
I think the default message could be improved to better convey the situation and intention behind marking issues as stale, without actually needing context about first-time contributors or anything like that.

Consider the default message:

> This issue has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.

This hits the hopeful submitter with two short negative statements, and then a very formal "thank you" which feels a bit like a business-like euphemism for "go away" when following an already negative message. There is no direct call to action, and an inexperienced submitter may interpret it as telling them that the issue will now be closed, while an experienced submitter may just be annoyed.

I think that just by offering the submitter something to do, the message takes a much friendlier tone:

> This issue has been automatically marked as stale because it has not had recent activity, and will be closed if no further activity occurs. If this issue was overlooked, forgotten, or should remain open for any other reason, please reply here to call attention to it and remove the stale status. Thank you for your contributions.

**Comment by waldyrious** [comment]
>  I believe the first-pass of stale-bot interacting with an issue should be prompting both the user and the maintainers for any more action

I fully agree with this, and I'm happy to say that it's actually already possible with the bot's current functionality -- it just isn't the default.

In the tldr-pages project we ended up [creating a configuration file](https://github.com/tldr-pages/tldr/pull/1360) that gets the bot from this:

> @probot-stale: "This issue has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.
> _Label: `wontfix`_

> (7 days later) @probot-stale closed #1234.

to this:

> @probot-stale: "Hi all! This thread has not had any recent activity.
> Are there any updates? Thanks!"
> _Label: `waiting`_

> (30 days later)
> @probot-stale: "Hi everyone.
> This thread is being closed as there was no response to the previous prompt.
> However, please leave a comment whenever you're ready to resume, so the thread can be reopened. Thanks again!"
> @probot-stale closed #1234.

Of note:
- the difference in days between the heads-up and the actual closing
- the friendlier tone of messages
- the less final-sounding label (cf. #74)
- the addressing of both contributors and maintainers
- the addition of a message when closing
- the suggestion of possibility of reopening

**Comment by waldyrious** [comment]
Oh, the irony... Yes, it's still relevant. Merging #76 would be a good way to make some progress, unless there are objections.

**Comment by bkeepers** [comment]
@waldyrious yes, that'd be great!

**Comment by dwijnand** [comment]
This looks to just be waiting for #162 to be reviewed and merged.

**Comment by dwijnand** [comment]
This looks to just be waiting for #162 to be reviewed and merged.

**Comment by Download** [comment]
I find it typical that the stale bot is pestering this very issue :)

Does the bot have access to the issue text? If so, we might add a feature that allows a user to comment something like:

```
blockedon #162 
```

to inform the stale bot that it should leave the issue alone until #162 is resolved.
